File: 프로그래머스/Lv2/타겟 넘버.cpp
#include <string>
#include <vector>

using namespace std;
int count=0;

void dfs(int Index, vector<int> numbers,int target,int total){
    if(Index>=numbers.size()){
        if(total==target) count++;
        return;
    }
    dfs(Index+1,numbers, target, total+numbers[Index]); 
    dfs(Index+1,numbers, target, total-numbers[Index]);
}

int solution(vector<int> numbers, int target) {
    count=0;
    dfs(0,numbers,target,0);
    return count;
}

File: 프로그래머스/Lv2/타겟 넘버.py
def solution(numbers, target):
    l=len(numbers)
    
    def dfs(sum,Index):
        nonlocal answer #중첩함수에서 바깥함수 변수 사용하기
        if(Index==l): 
            if(sum==target): answer+=1
            return
        
        dfs(sum+numbers[Index],Index+1)
        dfs(sum-numbers[Index],Index+1)
    
    answer=0
    dfs(0,0)
    
    return answer
